File: solutions/go/spiral-matrix/1/spiral_matrix.go
package spiralmatrix

func SpiralMatrix(size int) [][]int {
	top, bottom := 0, size - 1
    left, right := 0, size - 1
    result := make([][]int, size)
    for i := range result {
        result[i] = make([]int, size)
    }
    counter := 1
    for top <= bottom && left <= right {
        for i := left; i <= right; i++ {
            result[top][i] = counter
            counter++
        }
        top++
        for i := top; i <= bottom ; i++ {
            result[i][right] = counter
            counter++
        }
        right--
        if top <= bottom {
            for i := right; i >= left; i-- {
                result[bottom][i] = counter
                counter++
            }
            bottom--
        }
        if left <= right {
            for i := bottom; i >= top; i-- {
                result[i][left] = counter
                counter++
            }
            left++
        }
    }
    return result
}

File: solutions/csharp/spiral-matrix/1/SpiralMatrix.cs
public class SpiralMatrix
{
    public static int[,] GetMatrix(int size)
    {
        int[,] matrix = new int[size, size];

        int value = 1;
        int up = 0;
        int down = size - 1;
        int left = 0;
        int right = size - 1;
        while (value <= size*size)
        {
            for (int i = left; i <= right && value <= size * size; i++)
            {
                matrix[up, i] = value++;
            }
            up++;
            for (int i = up; i <= down && value <= size * size; i++)
            {
                matrix[i, right] = value++;
            }
            right--;
            for (int i = right; i >= left && value <= size * size; i--)
            {
                matrix[down, i] = value++;
            }
            down--;
            for (int i = down; i >= up && value <= size * size; i--)
            {
                matrix[i, left] = value++;
            }
            left++;
        }
        return matrix;
    }
}
